### PR TITLE
feat: install Windows csi proxy during cluster creation

### PR DESCRIPTION
**Pull request by marosset**
**Reason for Change**:

**Issue Fixed**:

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [x] includes documentation
- [x] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

**Comment by marosset** [comment]
To test this add the following to WindowsProfile:
      "enableCsiProxy": true,
      "csiProxyUrl": "https://marosseteastus2.blob.core.windows.net/pub/csi-proxy.tar"

(I'm updating clusterdefintion.md to document these fields right now)

File: parts/k8s/kuberneteswindowsfunctions.ps1 [hunk 1]
```diff
@@ -209,6 +209,7 @@ function Register-NodeResetScriptTask {
     Write-Log "Creating a startup task to run windowsnodereset.ps1"
 
     (Get-Content 'c:\AzureData\k8s\windowsnodereset.ps1') |
+    Foreach-Object { $_ -replace '{{CsiProxyEnabled}}', $global:EnableCsiProxy}
```
Comment (marosset) [inline parts/k8s/kuberneteswindowsfunctions.ps1]: this line needs a |

File: pkg/api/vlabs/validate.go [hunk 2]
```diff
@@ -1073,30 +1073,84 @@ func validateVMSS(o *OrchestratorProfile, isUpdate bool, storageProfile string)
...
+	if w == nil {
+		return errors.New("WindowsProfile is required when the cluster definition contains Windows agent pools")
+	}
+	if e := validate.Var(w.AdminUsername, "required"); e != nil {
+		return errors.New("WindowsProfile.AdminUsername is required, when agent pool specifies Windows")
+	}
+	if e := validate.Var(w.AdminPassword, "required"); e != nil {
+		return errors.New("WindowsProfile.AdminPassword is required, when agent pool specifies Windows")
+	}
+	if !validatePasswordComplexity(w.AdminUsername, w.AdminPassword) {
```
Comment (kkmsft) [inline pkg/api/vlabs/validate.go]: Are these related changes ?

File: pkg/api/vlabs/validate.go [hunk 3]
```diff
@@ -1073,30 +1073,84 @@ func validateVMSS(o *OrchestratorProfile, isUpdate bool, storageProfile string)
...
+	return nil
+}
+
+func validateCsiProxyWindowsProperties(w *WindowsProfile, k8sVersion string) error {
+	if w.IsCsiProxyEnabled() {
+		k8sSemVer, err := semver.Make(k8sVersion)
+		if err != nil {
+			return errors.Errorf("could not validate orchestrator version %s", k8sVersion)
+		}
+		minSemVer, err := semver.Make("1.18.0-alpha.2")
```
Comment (kkmsft) [inline pkg/api/vlabs/validate.go]: Should we switch to the beta version since we used that for one of the tests. Did you use the alpha for the VMSS one ?

Comment (marosset) [reply pkg/api/vlabs/validate.go]: I think i got it working with alpha.5 but i know beta.1 works so i'll update to beta.1

**Comment by andyzhangx** [comment]
@marosset  any progress on this? does this PR work now?

**Comment by marosset** [comment]
@andyzhangx - I'm going to pick this up again and try and get this merged today
This should work as is.

**Comment by marosset** [comment]
With the latest commit please use this package for testing https://kubernetesartifacts.azureedge.net/csi-proxy/master/binaries/csi-proxy.tar.gz

File: docs/topics/clusterdefinitions.md [hunk 4]
```diff
@@ -778,6 +778,8 @@ https://{keyvaultname}.vault.azure.net:443/secrets/{secretName}/{version}
 | ----------------------------- | -------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
 | adminUsername                 | yes      | Username for the Windows adminstrator account created on each Windows node                                                                                                                                                                      |
 | adminPassword                 | yes      | Password for the Windows adminstrator account created on each Windows node                                                                                                                                                                      |
+| csiProxyUrl                   | no       | Path to a package containing csi proxy binaries for Windows.                                                                                                                                                                                    |
```
[the remark this answers is not in the document]
Comment (marosset) [reply docs/topics/clusterdefinitions.md]: Once we have stable builds we'll default this URL.
We'll also want to add the csi proxy and related container images to the Windows VHD.

File: pkg/api/vlabs/validate.go (hunk 2 above)
Comment (marosset) [reply pkg/api/vlabs/validate.go]: Not related - I did some cleanup for how validation is performed for WindowsProfile here.

File: pkg/api/vlabs/validate.go (hunk 3 above)
Comment (marosset) [reply pkg/api/vlabs/validate.go]: updated

File: docs/topics/clusterdefinitions.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (marosset) [reply docs/topics/clusterdefinitions.md]: I added the URl of a package we can use for testing to the csi-proxy doc/markdown until we have official releases for the package.

File: pkg/api/const.go [hunk 5]
```diff
@@ -223,6 +223,8 @@ const (
 	DefaultMaximumLoadBalancerRuleCount = 250
 	// DefaultEnableAutomaticUpdates determines the aks-engine provided default for enabling automatic updates
 	DefaultEnableAutomaticUpdates = true
+	// DefaultEnableCsiProxyWindows determines if csi proxy should be enabled by default for Windows nodes
+	DefaultEnableCsiProxyWindows = false
```
Comment (jackfrancis) [inline pkg/api/const.go]: nit: s/DefaultEnableCsiProxyWindows/DefaultEnableCSIProxyWindows

File: pkg/api/types.go [hunk 6]
```diff
@@ -209,6 +209,8 @@ type CustomNodesDNS struct {
 type WindowsProfile struct {
 	AdminUsername             string            `json:"adminUsername"`
 	AdminPassword             string            `json:"adminPassword" conform:"redact"`
+	CsiProxyURL               string            `json:"csiProxyUrl"`
+	EnableCsiProxy            *bool             `json:"enableCsiProxy,omitempty"`
```
Comment (jackfrancis) [inline pkg/api/types.go]: nit: s/EnableCsiProxy/EnableCSIProxy and s/enableCsiProxy/enableCSIProxy

I know

File: pkg/api/types.go [hunk 7]
```diff
@@ -1669,6 +1671,14 @@ func (a *AgentPoolProfile) GetKubernetesLabels(rg string, deprecated bool) strin
 	return buf.String()
 }
 
+// IsCsiProxyEnabled returns true if csi proxy service should be enable for Windows nodes
+func (w *WindowsProfile) IsCsiProxyEnabled() bool {
```
Comment (jackfrancis) [inline pkg/api/types.go]: nit: s/IsCsiProxyEnabled/IsCSIProxyEnabled

File: pkg/api/vlabs/types.go [hunk 8]
```diff
@@ -161,6 +161,8 @@ type CustomNodesDNS struct {
 type WindowsProfile struct {
 	AdminUsername          string            `json:"adminUsername,omitempty"`
 	AdminPassword          string            `json:"adminPassword,omitempty"`
+	CsiProxyURL            string            `json:"csiProxyUrl"`
+	EnableCsiProxy         *bool             `json:"enableCsiProxy,omitempty"`
```
Comment (jackfrancis) [inline pkg/api/vlabs/types.go]: ditto above

File: pkg/api/vlabs/types.go [hunk 9]
```diff
@@ -883,6 +885,14 @@ func (l *LinuxProfile) HasCustomNodesDNS() bool {
 	return false
 }
 
+// IsCsiProxyEnabled returns true if csi proxy service should be enable for Windows nodes
+func (w *WindowsProfile) IsCsiProxyEnabled() bool {
```
Comment (jackfrancis) [inline pkg/api/vlabs/types.go]: ditto ditto

**Comment by marosset** [comment]
I'll update casing everywhere...

File: pkg/api/vlabs/types.go (hunk 9 above)
Comment (marosset) [reply pkg/api/vlabs/types.go]: Updating everywhere (except in the powershell code :))

File: pkg/api/vlabs/types.go [hunk 10]
```diff
@@ -883,6 +885,14 @@ func (l *LinuxProfile) HasCustomNodesDNS() bool {
 	return false
 }
 
+// IsCSIProxyEnabled returns true if CSI proxy service should be enable for Windows nodes
+func (w *WindowsProfile) IsCSIProxyEnabled() bool {
+	if w.EnableCSIProxy != nil {
+		return *w.EnableCSIProxy
+	}
+	return false // DO NOT HARDCODE
```
Comment (jackfrancis) [inline pkg/api/vlabs/types.go]: should this be here?

Comment (marosset) [reply pkg/api/vlabs/types.go]: probably not :P

File: pkg/api/const.go [hunk 11]
```diff
@@ -223,6 +223,8 @@ const (
 	DefaultMaximumLoadBalancerRuleCount = 250
 	// DefaultEnableAutomaticUpdates determines the aks-engine provided default for enabling automatic updates
 	DefaultEnableAutomaticUpdates = true
+	// DefaultEnableCSIProxyWindows determines if CSI proxy should be enabled by default for Windows nodes
+	DefaultEnableCSIProxyWindows = false
```
Comment (jackfrancis) [inline pkg/api/const.go]: you can move this to api/common/const.go if it'll be shared by api/vlabs

Comment (marosset) [reply pkg/api/const.go]: oh that works - let me do that instead of what I just did

Comment (marosset) [reply pkg/api/const.go]: I moved it to api/common/const.go

File: pkg/api/types.go [hunk 12]
```diff
@@ -1669,6 +1671,14 @@ func (a *AgentPoolProfile) GetKubernetesLabels(rg string, deprecated bool) strin
 	return buf.String()
 }
 
+// IsCSIProxyEnabled returns true if csi proxy service should be enable for Windows nodes
+func (w *WindowsProfile) IsCSIProxyEnabled() bool {
+	if w.EnableCSIProxy != nil {
+		return *w.EnableCSIProxy
+	}
+	return DefaultEnableCSIProxyWindows
```
Comment (jackfrancis) [inline pkg/api/types.go]: this needs to be `common.DefaultEnableCSIProxyWindows` (assuming `common` is a reference to the api/common package)

File: parts/k8s/kuberneteswindowssetup.ps1 [hunk 13]
```diff
@@ -378,6 +383,11 @@ try
             -KubeletNodeLabels $global:KubeletNodeLabels `
             -UseContainerD $useContainerD
 
+        if ($global:EnableCsiProxy) {
+            New-CsiProxyService -CsiProxyPackageUrl $global:CsiProxyUrl -KubeDir $global:KubeDir
+            # TODO: make kubelet depeond on csi-server-proxy service if enabled
```
Comment (kkmsft) [inline parts/k8s/kuberneteswindowssetup.ps1]: nit: depend

File: parts/k8s/kuberneteswindowssetup.ps1 [hunk 14]
```diff
@@ -378,6 +383,11 @@ try
             -KubeletNodeLabels $global:KubeletNodeLabels `
             -UseContainerD $useContainerD
 
+        if ($global:EnableCsiProxy) {
+            New-CsiProxyService -CsiProxyPackageUrl $global:CsiProxyUrl -KubeDir $global:KubeDir
```
Comment (kkmsft) [inline parts/k8s/kuberneteswindowssetup.ps1]: Could we please add a note in the README on where the csi proxy logs can be found in default setup.

File: parts/k8s/windowscsiproxyfunc.ps1 [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+    $tempdir = New-TemporaryDirectory
+    $binaryPackage = "$tempdir\csiproxy.tar"
+
+    DownloadFileOverHttp -Url $CsiProxyPackageUrl -DestinationPath $binaryPackage
+
+    tar -xzf $binaryPackage -C $tempdir
+    cp "$tempdir\build\server.exe" "$KubeDir\csi-proxy-server.exe"
+
+    del $tempdir -Recurse
```
Comment (kkmsft) [inline parts/k8s/windowscsiproxyfunc.ps1]: Would this happen whenever a new node is installed ?

File: parts/k8s/windowsnodereset.ps1 [hunk 16]
```diff
@@ -83,6 +89,12 @@ if ($global:NetworkPlugin -eq 'kubenet') {
 #
 # Start Services
 #
+
+if ($global:CsiProxyEnabled) {
+    Write-Log "Starting csi-proxy-server service"
```
Comment (kkmsft) [inline parts/k8s/windowsnodereset.ps1]: I was thinking we would also be introducing service dependancy ? Would the start-service do sync start or would it do a start asynchronously potentially causing race scenarios with start of kubelet ?

File: docs/topics/csi-proxy-windows.md [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+CSI Proxy for Windows enables support for CSI Plugins on Kubernetes nodes running Windows.
+
+More info can be found in the following places:
+
+- <https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20190714-windows-csi-support.md>
+- <https://github.com/kubernetes-csi/csi-proxy>
+
+## Requirements
+
+- CSI Proxy for Windows requires Kubernetes version 1.18.0 or greater.
```
Comment (kkmsft) [inline docs/topics/csi-proxy-windows.md]: Can we specify beta here ?

**Comment by mboersma** [comment]
@marosset our friendly acs-bot does not seem to have its `tide` config correct, so it merged this prematurely. I reverted the squashed commit in master and those responsible have been sacked.

**Comment by andyzhangx** [comment]
this PR could fix that issue: https://github.com/Azure/aks-engine/pull/2928
would you use that in the new PR? thanks.  @marosset

File: parts/k8s/windowsnodereset.ps1 (hunk 16 above)
Comment (marosset) [reply parts/k8s/windowsnodereset.ps1]: Oh yea - forgot we had discussed this.
I will add one.

**Comment by marosset** [comment]
whoops - i create a new PR this morning.
